hw/opentitan/ot_usbdev.c

```diff
@@ -0,0 +1,827 @@
+    REG_NAME_ENTRY(CONFIGIN_0),
+    REG_NAME_ENTRY(CONFIGIN_1),
+    REG_NAME_ENTRY(CONFIGIN_2),
+    REG_NAME_ENTRY(CONFIGIN_3),
+    REG_NAME_ENTRY(CONFIGIN_4),
+    REG_NAME_ENTRY(CONFIGIN_5),
+    REG_NAME_ENTRY(CONFIGIN_6),
+    REG_NAME_ENTRY(CONFIGIN_7),
+    REG_NAME_ENTRY(CONFIGIN_8),
+    REG_NAME_ENTRY(CONFIGIN_9),
+    REG_NAME_ENTRY(CONFIGIN_10),
+    REG_NAME_ENTRY(CONFIGIN_11),
+    REG_NAME_ENTRY(OUT_ISO),
+    REG_NAME_ENTRY(IN_ISO),
+    REG_NAME_ENTRY(OUT_DATA_TOGGLE),
+    REG_NAME_ENTRY(IN_DATA_TOGGLE),
+    REG_NAME_ENTRY(PHY_PINS_SENSE),
+    REG_NAME_ENTRY(PHY_PINS_DRIVE),
+    REG_NAME_ENTRY(PHY_CONFIG),
+    REG_NAME_ENTRY(WAKE_CONTROL),
+    REG_NAME_ENTRY(WAKE_EVENTS),
+    REG_NAME_ENTRY(FIFO_CTRL),
+    REG_NAME_ENTRY(COUNT_OUT),
+    REG_NAME_ENTRY(COUNT_IN),
+    REG_NAME_ENTRY(COUNT_NODATA_IN),
+    REG_NAME_ENTRY(COUNT_ERRORS),
+    /* clang-format on */
+};
+#undef REG_NAME_ENTRY
+
+/*
+ * State handling
+ */
+static void ot_usbdev_set_vbus_sense(OtUsbdevState *s, bool sense)
+{
+    s->vbus_sense = sense;
+    /* @todo need to update the link state and trigger some events here */
+}
+
+static void ot_usbdev_update_irqs(OtUsbdevState *s)
+{
+    uint32_t state_masked =
+        s->regs[R_USBDEV_INTR_STATE] & s->regs[R_USBDEV_INTR_ENABLE];
+
+    trace_ot_usbdev_irqs(s->ot_id, s->regs[R_USBDEV_INTR_STATE],
+                         s->regs[R_USBDEV_INTR_ENABLE], state_masked);
+
+    for (unsigned irq_index = 0; irq_index < USBDEV_INTR_NUM; irq_index++) {
+        bool level = (state_masked & (1U << irq_index)) != 0;
+        ibex_irq_set(&s->irqs[irq_index], level);
+    }
+}
+
+static void ot_usbdev_update_alerts(OtUsbdevState *s)
+{
+    uint32_t level = s->regs[R_ALERT_TEST];
+    s->regs[R_ALERT_TEST] = 0u;
+
+    /* @todo add other sources of alerts here */
+
+    if (level & R_ALERT_TEST_FATAL_FAULT_MASK) {
+        ibex_irq_set(&s->alert, 1);
+        ibex_irq_set(&s->alert, 0);
+    }
+}
+
+/*
+ * Register read/write handling
+ */
+
+static uint64_t ot_usbdev_read(void *opaque, hwaddr addr, unsigned size)
+{
+    OtUsbdevState *s = opaque;
+    (void)size;
+    uint32_t val32;
+
+    hwaddr reg = R32_OFF(addr);
+    switch (reg) {
+    case R_USBDEV_INTR_STATE:
+    case R_USBDEV_INTR_ENABLE:
+    case R_USBDEV_INTR_TEST:
+        val32 = s->regs[reg];
+        break;
+    case R_ALERT_TEST:
+        qemu_log_mask(LOG_GUEST_ERROR,
+                      "%s: %s Read to W/O register 0x%02" HWADDR_PRIx " (%s)\n",
+                      __func__, s->ot_id, addr, REG_NAME(reg));
+        val32 = 0;
+        break;
+    case R_USBCTRL:
+    case R_EP_OUT_ENABLE:
+    case R_EP_IN_ENABLE:
+    case R_USBSTAT:
+        /*
+         * For now, there is no host so we always always operate in
+         * vbus-override mode.
+         * @todo Report other statuses.
+         */
+        val32 = s->link_state << R_USBSTAT_LINK_STATE_SHIFT |
+                (uint32_t)s->vbus_sense << R_USBSTAT_SENSE_SHIFT;
```

Comment:
Maybe more idiomatic in QEMU is:

```suggestion
        val32 = FIELD_DP32(0u, USBSTAT, LINK_STATE, s->link_state);
        val32 = FIELD_DP32(val32, USBSTAT, SENSE, (uint32_t)s->vbus_sense);
```
But ultimately this is personal preference.